### PR TITLE
servo: Add an `Info.plist` on the Mac and opt into integrated graphics.

### PR DESCRIPTION
**Pull request by pcwalton**
Discrete GPUs cause power use problems and tend to perform badly
with WebRender.

See:
- https://developer.apple.com/library/mac/qa/qa1734/_index.html
- https://reverse.put.as/2013/05/28/gimmedebugah-how-to-embedded-a-info-plist-into-arbitrary-binaries/

r? @metajack

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/11646)

### CODE REVIEW

**Comment by KiChjang** [comment]
Fails tidy:
```
103.74s$ ./mach test-tidy --no-progress --all

Checking files for tidiness...

./components/servo/info_plist.rs:10: missing space after ,

./components/servo/info_plist.rs:10: missing space before =

./components/servo/info_plist.rs:10: missing space after =

./components/servo/info_plist.rs:12: missing space before -

./components/servo/info_plist.rs:16: missing space after /

./components/servo/info_plist.rs:17: missing space after /

./components/servo/info_plist.rs:18: missing space after /

./components/servo/info_plist.rs:19: missing space after /

./components/servo/info_plist.rs:20: missing space after /

./components/servo/info_plist.rs:21: missing space after /

./components/servo/info_plist.rs:22: missing space after /

./components/servo/info_plist.rs:23: missing space before /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:23: missing space after /

./components/servo/info_plist.rs:24: missing space after /

./components/servo/info_plist.rs:25: missing space after /

./components/servo/info_plist.rs:26: missing space after /

./components/servo/info_plist.rs:27: missing space after /

./components/servo/info_plist.rs:28: missing space after /

./components/servo/info_plist.rs:29: missing space after /

./components/servo/info_plist.rs:30: missing space after /

./components/servo/info_plist.rs:31: missing space after /
```

**Comment by metajack** [comment]
@pcwalton Is this integrated with the work on `./mach package`

@edunham might want to look at this too.

File: components/servo/info_plist.rs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+<plist version="1.0">
+<dict>
+       <key>CFBundleInfoDictionaryVersion</key>
+       <string>6.0</string>
+       <key>NSHumanReadableCopyright</key>
+       <string>Copyright &#169; 2016 The Servo Authors</string>
+       <key>CFBundlePackageType</key>
+       <string>APPL</string>
+       <key>CFBundleIconFile</key>
+       <string>../../resources/servo.png</string>
```
Comment (paulrouget) [inline components/servo/info_plist.rs]: This might not always work. Can you use `resources_dir_path()` for this?

Comment (pcwalton) [inline components/servo/info_plist.rs]: Oh, oops. This is an old version of the patch. Will fix ASAP. Do not merge.

File: components/servo/main.rs [hunk 2]
```diff
@@ -97,6 +101,8 @@ fn unregister_glutin_resize_handler(window: &Rc<app::window::Window>) {
     }
 }
 
+// Just as the name implies, this is a hack to prevent the linker from stripping our `info.plist`.
```
Comment (jdm) [inline components/servo/main.rs]: Weird place for this comment.

**Comment by pcwalton** [comment]
@metajack Updated. r?

I believe your `Info.plist` that you place in the `.app` bundle will override the one here, so it shouldn't cause any problems.

**Comment by metajack** [comment]
There are several problems here:

- The code unconditionally imports Cocoa, which may cause mach to fail and which is not documented. This should be optional just like our build length notification.
- It's not clear why you are adding info_plist.rs to the ignores for tidy.
- I don't want to see `info_plist::hack_to_prevent_linker_from_stripping_the_info_plist()` in `main()`. Can we hide this better? Maybe we need a `platform::init()` where this kind of stuff can go.
- The Info.plist file needs to be stored somewhere in tree and included `with include_str!`, so that it can be reused by the `./mach package` command. I don't mind it being linked in and on disk in a final package if they are the same, but we should at least not have two copies around or have to extract things from a build artifact.

On another note, is there seriously no way to tell rust not to optimize away a function? That is such an ugly hack that we should make sure a rust bug is filed for this and have a // TODO over that function that references that bug so that we know the criteria for when we can clean it up later.

**Comment by jdm** [comment]
There's [test::black_box](https://github.com/rust-lang/rust/blob/0ec321f7b541fcbfbf20286beb497e6d9d3352b2/src/libtest/lib.rs#L1196).

**Comment by pcwalton** [comment]
> It's not clear why you are adding info_plist.rs to the ignores for tidy.

This is because tidy doesn't realize that `__TEXT,__info_plist` is inside a quoted string and will fail telling me to insert a space after the comma. That breaks because Mach-O (I believe) requires no space after the comma.

**Comment by metajack** [comment]
@pcwalton Fair, but let's file a bug against tidy and annotate the reason in the ignores list.

**Comment by edunham** [comment]
@metajack `mach package` currently only supports Linux and Android. 

I haven't worked with OSX packaging much, but I don't see anything obviously wrong in these changes.

**Comment by pcwalton** [comment]
@jdm Sadly `test::black_box` only [prevents LLVM's optimizer from optimizing out the symbol](https://doc.rust-lang.org/1.1.0/src/test/lib.rs.html#1054-1059), which doesn't help us in this case because it's the *linker* that strips out the symbol.

**Comment by pcwalton** [comment]
Comments addressed.

r? @metajack

**Comment by metajack** [comment]
Reviewable has totally failed me on this patch. Sigh.

**Comment by metajack** [comment]
@bors-servo r+

Thanks for these updates.

**Pull request merged**